jms/1.8/modules/ROOT/pages/jms-troubleshooting.adoc

```diff
@@ -0,0 +1,125 @@
+
+== Troubleshoot Acknowledgment Issues
+
+There are several ways an application setup can be inconsistent in terms of message acknowledgment.
+You can configure the session acknowledgment for an application as:
+
+* *IMMEDIATE*
+* *AUTO*
+* *MANUAL*
+* *DUPS_OK*
+
+When you set the session acknowledgment to *IMMEDIATE*, the message gets acknowledged as soon as it's received. Do not mix *IMMEDIATE* acknowledgment with *Ack* or *Recovery session* operations.
+
+When you set the session acknowledgment to *AUTO*, the *On New Message* source takes control of the acknowledgment when the event finishes, or the session recovers when an error is thrown while processing the event. Adding a manual acknowledgment or session recover in an application with *AUTO* acknowledgment leads to duplicate operations on a channel.
+
+When you set the session acknowledgment to either *MANUAL* or *DUPS_OK*, check that every path of the flow execution gets a message acknowledgment or a session recover. Otherwise, a message can leak and lead to an unresponsive consumer when the max depth of the unit of work is reached.
+
+To resolve this issue:
+
+Take a detailed look at the flow design and be sure that every message is acknowledged or the session recovers.
+
+== Troubleshoot JMS Stopped Consuming Messages
+
+Usually when JMS stopped consuming messages is related to a large number of unacknowledged messages. When your  Mule app that has an *On New Message* source or *Consume* operation with manual acknowledgement of messages, and not every path of the message has the proper acknowledgement or session recovery configuration, the Mule app can lead to an unresponsive consumer.
+
+To resolve this error:
+
+Check the depth of unacknowledged messages on the broker side and review the application logic.
+
+== Troubleshoot Redeployment Errors When Using Binding Mode
+
+When using binding mode for connections, the driver loads a JNI resource. Because the JNI resource can be loaded only once, when the application is redeployed, an error is thrown.
+
+To resolve this error:
+
+Add the driver dependency at the domain level, which enables application redeployment.
+
+== Troubleshoot HA Environments and Durable Subscriptions
+
+The configuration of HA (High Availability) systems consists of two components:
+
+* Primary node +
+Is the main node that provides continuous service.
+* Secondary node +
+Is a clone of the primary node and provides the same service.
+
+Both nodes must be identical so the end-user is not aware of the system crash and can continue to use the system transparently.
+
+However, when you work with durable subscriptions, only one node initialize a connection while the other node fails to do it. In these scenarios, the nodes are not clones because one node has a connection while the other one doesn't. The current design of Mule running in HA clusters aims that all nodes initialize connections to ensure that they are capable of facing a failure of the primary node.
```

Comment:
```suggestion
However, when you work with durable subscriptions, only one node initializes a connection while the other node fails to do it. In these scenarios, the nodes are not clones because one node has a connection while the other one doesn't. The current design of Mule running in HA clusters aims that all nodes initialize connections to ensure that they are capable of facing a failure of the primary node.
```